plutus-tx/src/PlutusTx/Optimize/SpaceTime.hs

```diff
@@ -1,37 +1,99 @@
+{-| Given @n@, and the step function for a recursive function, peel @n@ layers
 off of the recursion.
 
-For example @peel 2 (\f xs -> case xs of [] -> 0; (_:ys) -> 1 + f ys)@ yields the
-equivalence of the following function:
+For example @peel 3 (\self -> [[| \case [] -> 0; _ : ys -> 1 + self ys||])@
+yields the equivalence of the following function:
 
 @
-   lengthPeeled :: [a] -> a
-   lengthPeeled xs = case xs of
-     [] -> 0
-     y:ys -> 1 + case ys of
-       [] -> 0
-       z:zs -> 1 + case zs of
-         [] -> 0
-         w:ws -> 1 + length ws
+  lengthPeeled :: [a] -> a
+  lengthPeeled xs =
+    case xs of                     -- first recursion step
+      []     -> 0
+      _ : ys -> 1 +
+        case ys of                 -- second recursion step
+          []     -> 0
+          _ : zs -> 1 +
+            case zs of             -- third recursion step
+              []     -> 0
+              _ : ws -> 1 +
+                ( fix \self qs ->  -- rest of recursion steps in a tight loop
+                    case qs of
+                      []     -> 0
+                      _ : ts -> 1 + self ts
+                ) ws
 @
 
-where @length@ is the regular recursive definition.
 -}
 peel
   :: forall a b
    . Int
+  -- ^ How many recursion steps to move outside of the recursion loop.
   -> (TH.SpliceQ (a -> b) -> TH.SpliceQ (a -> b))
+  {- ^ Function that given a continuation splice returns
+  a splice representing a single recursion step calling this continuation.
+  -}
   -> TH.SpliceQ (a -> b)
-peel 0 f = [||fix (\g -> $$(f [||g||]))||]
+peel 0 f = [||fix \self -> $$(f [||self||])||]
 peel n f
   | n > 0 = f (peel (n - 1) f)
   | otherwise = error $ "PlutusTx.Optimize.SpaceTime.peel: negative n: " <> show n
+
+{-| Given @n@, and the step function for a recursive function,
+    unroll recursion @n@ layers at a time
+
+For example @unroll 3 (\self -> [|| \case [] -> 0; _ : ys -> 1 + self ys ||])@
+yields the equivalence of the following function:
+
+@
+  lengthUnrolled :: [a] -> a
+  lengthUnrolled =
+    fix \self xs ->                   -- beginning of the recursion "loop"
+      case xs of                      -- first recursion step
+        []     -> 0
+        _ : ys -> 1 +
+          case ys of                  -- second recursion step
+            []     -> 0
+            _ : zs -> 1 +
+              case zs of              -- third recursion step
+                []     -> 0
+                _ : ws -> 1 + self ws -- end of the "loop"
+
+@
+
+-}
+unroll
+  :: forall a b
+   . Int
+  -- ^ How many recursion steps to perform inside the recursion loop.
+  -> (TH.SpliceQ (a -> b) -> TH.SpliceQ (a -> b))
+  {- ^ Function that given a continuation splice returns
+  a splice representing a single recursion step calling this continuation.
+  -}
+  -> TH.SpliceQ (a -> b)
+unroll n f = [||fix \self -> $$(unrollBound n f [||self||])||]
+
+unrollBound
```

Comment:
This just composes the function with itself `n` times, right? So perhaps call it `nTimes`? There's already a [similar function](https://hackage.haskell.org/package/ghc-9.12.1/docs/GHC-Utils-Misc.html#v:nTimes) (non-TH) in `GHC.Utils.Misc`.

Reply:
Thanks, updated to code.